drivers/store/redis/store.go

```diff
@@ -0,0 +1,288 @@
+	pipe := rtx.Pipeline()
+	value := pipe.Get(key)
+	expire := pipe.PTTL(key)
+
+	_, err := pipe.Exec()
+	if err != nil && err != libredis.Nil {
+		return 0, 0, err
+	}
+
+	count, err := value.Int64()
+	if err != nil && err != libredis.Nil {
+		return 0, 0, err
+	}
+
+	ttl, err := expire.Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return count, ttl, nil
+}
+
+// doSetValue will execute setValue with a retry mecanism (optimistic locking) until store.MaxRetry is reached.
+func (store *Store) doSetValue(rtx *libredis.Tx, key string, expiration time.Duration) (bool, error) {
+	for i := 0; i < store.MaxRetry; i++ {
+		created, err := store.setValue(rtx, key, expiration)
+		if err == nil {
+			return created, nil
+		}
+	}
+	return false, errors.New("retry limit exceeded")
+}
+
+// setValue will try to initialize a new counter if given key doesn't exists.
+func (store *Store) setValue(rtx *libredis.Tx, key string, expiration time.Duration) (bool, error) {
+	pipe := rtx.Pipeline()
+	value := pipe.SetNX(key, 1, expiration)
+
+	_, err := pipe.Exec()
+	if err != nil {
+		return false, err
+	}
+
+	created, err := value.Result()
+	if err != nil {
+		return false, err
+	}
+
+	return created, nil
+}
+
+// doUpdateValue will execute setValue with a retry mecanism (optimistic locking) until store.MaxRetry is reached.
+func (store *Store) doUpdateValue(rtx *libredis.Tx, key string,
+	expiration time.Duration) (int64, time.Duration, error) {
+
+	for i := 0; i < store.MaxRetry; i++ {
+		count, ttl, err := store.updateValue(rtx, key, expiration)
+		if err == nil {
+			return count, ttl, nil
+		}
+
+		// If ttl is negative and there is an error, do not retry an update.
+		if ttl < 0 {
+			return 0, 0, err
+		}
+	}
+	return 0, 0, errors.New("retry limit exceeded")
+}
+
+// updateValue will try to increment the counter identified by given key.
+func (store *Store) updateValue(rtx *libredis.Tx, key string, expiration time.Duration) (int64, time.Duration, error) {
+	pipe := rtx.Pipeline()
+	value := pipe.Incr(key)
+	expire := pipe.PTTL(key)
+
+	_, err := pipe.Exec()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	count, err := value.Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	ttl, err := expire.Result()
+	if err != nil {
+		return 0, 0, err
+	}
+
+	// If ttl is negative, we have to define key expiration.
+	if ttl < 0 {
+		expire := rtx.Expire(key, expiration)
+
+		ok, err := expire.Result()
+		if err != nil {
+			return count, ttl, err
+		}
+
+		if !ok {
+			return count, ttl, errors.New("cannot configure timeout on key")
+		}
+	}
+
+	return count, ttl, nil
+
+}
+
+// ping checks if redis is alive.
+func (store *Store) ping() (bool, error) {
+	cmd := store.client.Ping()
+
+	pong, err := cmd.Result()
+	if err != nil {
+		return false, errors.Wrap(err, "limiter: cannot ping redis server")
+	}
+
+	return (pong == "PONG"), nil
+}
+
+func (store *Store) getContextFromState(now time.Time, rate limiter.Rate,
```

Comment:
same

Reply:
it seems this method is the same exact code as ``memory.store. getContextFromState``, let's re-use it as a static utility